### PR TITLE
feat(mint): implement ADR 0059 public mint mode

### PR DESCRIPTION
**Pull request by ifireball**
## Summary
- Add public mint mode when `ALLOWED_ORGS` is `*`: any org may mint, but `job_workflow_ref` must come from `fullsend-ai/fullsend/.github/workflows/` only
- Keep tight mode unchanged (explicit org list, legacy `.fullsend`/per-repo provenance, basename allowlist)
- Update provisioner enrollment (`EnsureOrgInMint` succeeds without config changes, per-repo WIF/unenroll errors) and CLI messaging for public mints
- Add `mint deploy --public` to bootstrap GCF-hosted public mints with aligned `ALLOWED_ORGS=*` and permissive WIF CEL (`assertion.repository_owner != ''`); bidirectional mode guards block tight↔public conversion on redeploy
- Sync embedded GCF mint sources and update infrastructure docs

## Deploying a public GCF mint

```bash
fullsend mint deploy --project MY_PROJECT --pem-dir ./pems --public
```

- First-time bootstrap still requires `--pem-dir` (shared role PEMs/app IDs)
- Redeploy/upgrade must match mode: `--public` for public mints, omit for tight mints
- Standalone JWKS mints (`cmd/mint`) are unchanged — no WIF provisioning needed

## Test plan
- [x] `go test ./internal/mintcore/... ./internal/dispatch/gcf/... ./internal/cli/...`
- [x] `hack/lint-mint-embed-sync`
- [ ] Manual: `mint deploy --public` provisions permissive WIF + `ALLOWED_ORGS=*`; STS accepts any org with upstream workflow provenance
- [ ] Manual: redeploy guards reject mode mismatch (`--public` into tight / tight into public)

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: docs/guides/infrastructure/infrastructure-reference.md [hunk 1]
```diff
@@ -90,18 +92,33 @@ Custom roles can be registered via the standalone mint's `CUSTOM_ROLE_PERMISSION
...
+
+**Tight mint** (default): explicit comma-separated org list (no `*`).
+
+- **ALLOWED_ORGS**: Only listed orgs may mint tokens
+- **ALLOWED_WORKFLOW_FILES**: Fail-closed allowlist of workflow filenames (use `*` to allow any basename)
+- **job_workflow_ref validation**: `.fullsend` config repo, `fullsend-ai/fullsend` upstream reusables, or registered per-repo workflows (`PER_REPO_WIF_REPOS`)
 - **PER_REPO_WIF_REPOS**: Repos using dedicated WIF providers (repo-scoped isolation)
-- **Minimum permissions**: Tokens are scoped to the role's minimum permission set, not the App's full permissions
+
+**Public mint**: `ALLOWED_ORGS` contains `*`.
```
Comment (rh-hemartin) [inline docs/guides/infrastructure/infrastructure-reference.md]: Contains? I would say `ALLOWED_ORGS` is `*`.

File: docs/guides/infrastructure/mint-administration.md [hunk 2]
```diff
@@ -262,6 +262,12 @@ Prior versions of `mint enroll` accepted `--app-set`, `--role-app-ids`, `--roles
 
 Role PEM secrets and `ROLE_APP_IDS` must already exist on the mint, created during `mint deploy --pem-dir` or `mint add-role`. Enrollment does not create, copy, or modify PEM secrets or app ID mappings.
 
+### Public mint mode
+
+When the mint is configured with `ALLOWED_ORGS=*` ([ADR 0059](../../ADRs/0059-public-mint-mode-with-wildcard-allowlists.md)), `mint enroll` is a no-op for org registration — all orgs are already allowed. Per-repo WIF registration (`mint enroll owner/repo`) is not supported in public mode; per-repo installs use the default WIF provider and upstream reusable workflows.
```
Comment (rh-hemartin) [inline docs/guides/infrastructure/mint-administration.md]: What does it mean that is `no-op`? Can we say "it fails", "does not work", "leaves early", or something like that?

**Review by rh-hemartin — APPROVED** [review]

File: docs/guides/infrastructure/mint-administration.md (hunk 2 above)
Comment (ifireball) [reply docs/guides/infrastructure/mint-administration.md]: its not the the enrolment fails in the sense that the user will not be able to work, its that its not needed, so I think its better to define it as passing but doing nothing rather then failing. This will make it easier to e.g. write a script that works in both modes.

File: docs/guides/infrastructure/infrastructure-reference.md (hunk 1 above)
Comment (ifireball) [reply docs/guides/infrastructure/infrastructure-reference.md]: Updated to say `ALLOWED_ORGS` is `*` rather than "contains *".

File: docs/guides/infrastructure/mint-administration.md (hunk 2 above)
Comment (ifireball) [reply docs/guides/infrastructure/mint-administration.md]: Clarified in docs: `mint enroll <org>` succeeds without modifying mint configuration (per ifireball's note — scripts can call enroll in both modes without branching). Dropped "no-op" jargon.

Comment (rh-hemartin) [reply docs/guides/infrastructure/mint-administration.md]: Thanks!

File: docs/guides/infrastructure/infrastructure-reference.md (hunk 1 above)
Comment (rh-hemartin) [reply docs/guides/infrastructure/infrastructure-reference.md]: Thanks!

File: internal/mintcore/claims.go [hunk 3]
```diff
@@ -56,8 +56,27 @@ func (a Audience) Contains(aud string) bool {
...
 func ValidateOrgAllowed(org string, allowedOrgs []string) error {
+	if org == "" {
+		return fmt.Errorf("missing repository_owner claim")
+	}
+	if IsPublicMint(allowedOrgs) {
+		return nil
+	}
 	for _, entry := range allowedOrgs {
 		if strings.EqualFold(entry, org) {
 			return nil
```
[the remark this answers is not in the document]
Comment (ifireball) [reply internal/mintcore/claims.go:96]: Valid observation about the STS/WIF layer, but intentionally out of scope for this PR per [ADR 0059](https://github.com/fullsend-ai/fullsend/blob/main/docs/ADRs/0059-public-mint-mode-with-wildcard-allowlists.md): permissive WIF provider provisioning (CEL, pools, IAM) is deferred to a future mint infrastructure ADR. This PR implements the mintcore authorization layer (`ALLOWED_ORGS=*`, upstream-only workflow provenance) and enrollment CLI behavior; operators must pair `ALLOWED_ORGS=*` with a permissive `WIF_PROVIDER_NAME` when adopting public mode.

Added an explicit note in `infrastructure-reference.md` that setting `ALLOWED_ORGS=*` without a matching permissive provider will fail STS for orgs outside the provider condition, and that `mint deploy`/`mint enroll` still only provision tight-mode WIF today.

[the remark this answers is not in the document]
Comment (ifireball) [reply internal/mintcore/claims.go:96]: Addressed in e2e5230: `mint deploy --public` now provisions `ALLOWED_ORGS=*` and permissive WIF CEL (`assertion.repository_owner != ''`) together on fresh GCF bootstrap, with bidirectional redeploy guards (tight↔public conversion rejected). Docs updated in `docs/cli/mint.md`, `mint-administration.md`, and `infrastructure-reference.md`.

**Comment by ifireball** [comment]
Follow-up commits on this branch:

- **e2e5230** — `mint deploy --public` bootstraps GCF public mints with `ALLOWED_ORGS=*` + permissive WIF CEL; bidirectional redeploy guards reject tight↔public mode conversion
- **89cb3d19** — update `infrastructure-reference.md` to document `--public` WIF provisioning (replaces the earlier "mint deploy only provisions tight-mode WIF" note)

PR description updated accordingly. Replied on the Qodo WIF/STS thread — that concern is addressed for the GCF deploy path via `--public`.

**Comment by ifireball** [comment]
Babysit update (27d5fcde):

- **CI test**: fixed `gofmt` on `provisioner_test.go`
- **CI behaviour**: merged upstream `main` (adds `make behaviour-test` + e2e behaviour job from #2793)
- **Review**: `runMintUnenrollOrg` now exits early in public mode (before dry-run/confirm), matching enroll/repo handling

All checks green on latest push.

**Comment by ifireball** [comment]
## Review nits — b0103225

Addressing [fullsend-ai-review low/optional findings](https://github.com/fullsend-ai/fullsend/pull/2773#issuecomment-4842316538) that do not contradict ADR 0059.

### Fixed

| Finding | Resolution |
|---------|------------|
| **Error suppression** (`mint.go` enroll/unenroll org paths) | `GetServiceTrafficEnvVars` errors are now propagated instead of discarded (`trafficEnv, _ := …`). |
| **Duplicate `parseAllowedOrgs`** | Added `mintcore.ParseAllowedOrgs`; CLI and provisioner both use it. |
| **Doc ambiguity** (`mint-administration.md` public mode) | Clarified: `mint enroll` exits 0 in both modes; only tight mode updates `ALLOWED_ORGS`/WIF. |
| **`isPublicMintEnv` naming** | Renamed to `isPublicMintAllowedOrgs`. |
| **`mintValidationStepDone` naming** | Renamed to `mintValidationMessage` (CLI + `admin.go`). |
| **`mintPublicDiscoveryClient` test helper** | Renamed to `publicMintDiscoveryClient`. |
| **Empty workflow filename edge case** | Public-mode `ValidateWorkflowRef` now rejects refs with an empty basename after `.github/workflows/`. |

### Intentionally not fixed (ADR / scope)

| Finding | Reason |
|---------|--------|
| **Stale getting-started note** (`docs/guides/getting-started/README.md`) | Public mint remains user-facing “in progress” until a hosted public mint is deployed at `mint.fullsend.sh` (may need a new implementation). |
| **Fail-open `ALLOWED_ORGS=*`** | ADR 0059 Decision Point 1 — no secondary mode flag by design. |
| **Qodo guide path / `ghs_` test token** | Out of scope / existing repo convention. |

### Medium findings (separate from this pass)

- **`runMintUnenrollRepo` public-mode guard** — not in the low/optional set; can follow up if desired.
- **`EnsureOrgInWIFCondition` public-mode guard** — provisioner hardening; CLI enroll paths already return early in public mode.
- **ADR WIF permissiveness documentation** — partially covered by `mint deploy --public` + infrastructure docs from prior commits.

All existing inline review threads were already resolved; no new threads to resolve on this commit.

**Comment by ifireball** [comment]
## `runMintUnenrollRepo` public-mode guard — 4668f45c

Added the same early return used by `enroll org/repo` and `unenroll org`: after mint discovery, read `ALLOWED_ORGS` and exit successfully before dry-run, confirmation, or WIF changes when `ALLOWED_ORGS=*`. Covered by `TestRunMintUnenrollRepo_PublicMode`.

---

## Remaining medium findings — intentionally deferred

Inline review threads on this PR are **all resolved** (docs wording, enroll no-op, Qodo WIF/STS — last addressed via `mint deploy --public`). The items below only appear in [fullsend-ai-review bot comments](https://github.com/fullsend-ai/fullsend/pull/2773#issuecomment-4842316538), not as open threads.

### `EnsureOrgInWIFCondition` could overwrite permissive WIF CEL

**Not fixing in this PR.** CLI enroll paths already return before calling WIF updates in public mode. Hardening the provisioner API (guard all WIF mutators when `ALLOWED_ORGS=*`) is defense-in-depth for direct/non-CLI callers and deserves a small focused follow-up rather than expanding this PR.

### Permissive WIF (`assertion.repository_owner != ''`) documentation

**Not fixing in this PR.** ADR 0059’s layered model: mintcore is the authorization gate in public mode; WIF is intentionally permissive once paired via `mint deploy --public`. Already documented in `mint-administration.md`, `infrastructure-reference.md`, and `docs/cli/mint.md`. Additional ADR prose is optional polish.

### Stale getting-started “in progress” note

**Not fixing in this PR.** Public mint remains user-facing “in progress” until a hosted public mint is live at `mint.fullsend.sh` — self-hosted `mint deploy --public` does not change that end-user story yet.
